### PR TITLE
verify script to ensure publishing bot rules and godeps are in sync

### PR DESCRIPTION
**Pull request by dims**
Change-Id: Idbad862a797916cc4408ea0099c30d0557044907

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
```
[dims@dims-mac 12:10] ~/go/src/k8s.io/kubernetes ⟩ time hack/verify-publishing-bot.py > /dev/null

real	0m0.585s
user	0m0.390s
sys	0m0.128s
```

**Comment by dims** [comment]
/assign @sttts @nikhita 

Stefan, Nikhita,
this checks the rules.yaml against the godeps.json in the staging/ directory to make sure that all dependencies currently in the godeps.json are also present in rules.yaml (NOT the other way around since rules.yaml can contain more than what's in godeps.json).

**Comment by sttts** [comment]
Awesome!!!

**Comment by sttts** [comment]
Does it check both way, i.e. that the rules are comlete **and** correct, i.e. actually necessary?

**Comment by sttts** [comment]
It works on all branches?

**Comment by dims** [comment]
step 1: godeps.json is calculated to have just the minimum stuff right? we have scripts to verify that
step 2: this script will make sure everything in godeps.json is in rules.yaml

So yes, everything that is needed for the code to compile will make it. the code-generator stuff seems to be added only in rules.yaml (and is not present in godeps.json). so that would be the additional stuff in rules.yaml that is not in godeps.json

**Comment by dims** [comment]
since it validates rules.yaml with the godeps.json in that specific repository (just the "master" definitions in the rules.yaml) it should not fail when this script is run in say the release-1.14 branch

**Comment by nikhita** [comment]
@dims :tada: :tada: :tada: 

>  the code-generator stuff seems to be added only in rules.yaml (and is not present in godeps.json). so that would be the additional stuff in rules.yaml that is not in godeps.json

Can we look at `required-packages` and make sure that if something in `required-packages` doesn't exist in `godeps.json` it is still present in `dependencies`?

I'm gonna let someone super proficient in python review the actual code (which is not me) but overall the script looks good (it still needs to be invoked, I guess?)

And...iirc we had added `component-base` in rules because repos _would_ depend on it the future...maybe we should just remove `component-base` from the rules for now and add it back when it's needed? @dims I'm guessing the script catches this case?

**Comment by dims** [comment]
it's already invoked @nikhita look at the logs from `pull-kubernetes-verify`

yes, we can add more stuff over time for sure (like the `required-packages`). 

i am ok with leaving component-base as-is. (yes, we can check if all the repos that are staged are listed in rules.yaml too!)

**Comment by nikhita** [comment]
:+1: :100:

**Comment by dims** [comment]
@nikhita i tightened things further, we'll yell out load if the rules.yaml is missing something that has a Godeps.json or if the rules.yaml has a typo or something that does not have a corresponding Godeps.json.

Can we please check this in and iterate? we could add a good first item label for the following issue

> Can we look at required-packages and make sure that if something in required-packages doesn't exist in godeps.json it is still present in dependencies?

**Comment by spiffxp** [comment]
/milestone v1.14
I think this will help us with the release. Leaving lgtm for someone else unless I get more time later

File: hack/verify-publishing-bot.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        if branch["name"] != "master":
+            raise Exception("cannot find master branch for destination %s" % rule["destination"])
+        if branch["source"]["branch"] != "master":
+            raise Exception("cannot find master source branch for destination %s" % rule["destination"])
+
+        print("processing : %s" % rule["destination"])
+        if rule["destination"] not in godep_dependencies:
+            raise Exception("missing Godeps.json for %s" % rule["destination"])
+        processed_repos.append(rule["destination"])
+        for dep in set():
```
Comment (spiffxp) [inline hack/verify-publishing-bot.py]: I'm ok for merge-and-iterate with comments to be added later, except for this part here.  Is this an intentionally dead loop?  If so, a comment explaining why would be good for newcomers

Comment (dims) [reply hack/verify-publishing-bot.py]: nice catch! my bad, checked in a stale version

**Review by nikhita — APPROVED** [review]
:rocket: 

/lgtm
/sig release

**Pull request merged**